cpp/include/raft/neighbors/detail/cagra/graph_core.cuh

```diff
@@ -463,94 +1212,113 @@ void optimize(raft::resources const& res,
+      raft::copy(d_dest_nodes_ptr, dest_nodes_ptr, graph_size, resource::get_cuda_stream(res));
 
       dim3 threads(256, 1, 1);
       dim3 blocks(1024, 1, 1);
       kern_make_rev_graph<<<blocks, threads, 0, resource::get_cuda_stream(res)>>>(
-        d_dest_nodes.data_handle(),
-        d_rev_graph.data_handle(),
-        d_rev_graph_count.data_handle(),
+        d_dest_nodes_ptr,
+        d_rev_graph_ptr,
+        d_rev_graph_count_ptr,
         graph_size,
-        output_graph_degree);
+        output_graph_degree,
+        (k % 2));
       RAFT_LOG_DEBUG("# Making reverse graph on GPUs: %lu / %u    \r", k, output_graph_degree);
     }
 
     resource::sync_stream(res);
     RAFT_LOG_DEBUG("\n");
 
     if (d_rev_graph.allocated_memory()) {
-      raft::copy(rev_graph.data_handle(),
-                 d_rev_graph.data_handle(),
-                 graph_size * output_graph_degree,
+      raft::copy(rev_graph_ptr,
+                 d_rev_graph_ptr,
+                 (size_t)graph_size * output_graph_degree,
                  resource::get_cuda_stream(res));
     }
-    raft::copy(rev_graph_count.data_handle(),
-               d_rev_graph_count.data_handle(),
-               graph_size,
-               resource::get_cuda_stream(res));
+    raft::copy(
+      rev_graph_count_ptr, d_rev_graph_count_ptr, graph_size, resource::get_cuda_stream(res));
 
+    resource::sync_stream(res);
     const double time_make_end = cur_time();
     RAFT_LOG_DEBUG("# Making reverse graph time: %.1lf sec", time_make_end - time_make_start);
   }
 
   {
     //
-    // Replace some edges with reverse edges
+    // Merge forward and reverse edges
     //
     const double time_replace_start = cur_time();
-
-    const uint64_t num_protected_edges = output_graph_degree / 2;
-    RAFT_LOG_DEBUG("# num_protected_edges: %lu", num_protected_edges);
-
-    constexpr int _omp_chunk = 1024;
-#pragma omp parallel for schedule(dynamic, _omp_chunk)
-    for (uint64_t j = 0; j < graph_size; j++) {
-      uint64_t k = std::min(rev_graph_count.data_handle()[j], output_graph_degree);
-      while (k) {
-        k--;
-        uint64_t i = rev_graph.data_handle()[k + (output_graph_degree * j)];
-
-        uint64_t pos =
-          pos_in_array<IdxT>(i, output_graph_ptr + (output_graph_degree * j), output_graph_degree);
-        if (pos < num_protected_edges) { continue; }
-        uint64_t num_shift = pos - num_protected_edges;
-        if (pos == output_graph_degree) {
-          num_shift = output_graph_degree - num_protected_edges - 1;
+#pragma omp parallel for
+    for (uint64_t i = 0; i < graph_size; i++) {
+      auto my_fwd_graph = pruned_graph_ptr + (output_graph_degree * i);
+      auto my_rev_graph = rev_graph_ptr + (output_graph_degree * i);
+      auto my_out_graph = output_graph_ptr + (output_graph_degree * i);
+      uint32_t kf       = 0;
+      uint32_t kr       = 0;
+      uint32_t k        = mst_graph_num_edges_ptr[i];
```

Comment:
The code below keeps `k` edges from the MST, and fills the rest from the forward and reverse graph. Typically how large is `k` compared to `output_graph_degree`?